adapters/play/src/main/scala/caliban/PlayJson.scala

```diff
@@ -0,0 +1,67 @@
+package caliban
+
+import caliban.interop.play.json.parsingException
+import play.api.libs.json.{ JsObject, JsString, JsValue, Json }
+import play.api.http.Writeable
+import play.api.libs.json.Writes
+
+import scala.util.Try
+
+trait PlayJson {
+
+  implicit def writableGraphQLResponse[E](implicit wr: Writes[GraphQLResponse[E]]): Writeable[GraphQLResponse[E]] =
+    Writeable.writeableOf_JsValue.map(wr.writes)
+
+  private def parseJson(s: String): Try[JsValue] =
+    Try(Json.parse(s))
+
+  def parseHttpRequest(
+    query: String,
+    op: Option[String],
+    vars: Option[String],
+    exts: Option[String]
+  ): Either[Throwable, GraphQLRequest] = {
+    val variablesJs  = vars.flatMap(parseJson(_).toOption)
+    val extensionsJs = exts.flatMap(parseJson(_).toOption)
+    val fields = List("query" -> JsString(query)) ++
+      op.map(o => "operationName"         -> JsString(o)) ++
+      variablesJs.map(js => "variables"   -> js) ++
+      extensionsJs.map(js => "extensions" -> js)
+    JsObject(fields)
+      .validate[GraphQLRequest]
+      .asEither
+      .left
+      .map(parsingException)
+  }
+
+  def encodeGraphQLResponse(r: GraphQLResponse[Any]): String = Json.toJson(r).toString()
+
+  def parseWSMessage(text: String): Either[Throwable, PlayWSMessage] =
+    parseJson(text).toEither.map { json =>
+      PlayWSMessage(
+        (json \ "id").validate[String].getOrElse(""),
+        (json \ "type").validate[String].getOrElse(""),
+        (json \ "payload").validate[JsObject].asOpt
+      )
+    }
+
+  def encodeWSResponse[E](id: String, data: ResponseValue, errors: List[E]): String =
+    Json.stringify(
+      Json
+        .obj(
+          "id"      -> id,
+          "type"    -> "data",
+          "payload" -> GraphQLResponse(data, errors)
+        )
+    )
+
+  def encodeWSError(id: String, error: Throwable): String =
+    Json.stringify(
+      Json
+        .obj(
+          "id"      -> id,
+          "type"    -> "complete",
+          "payload" -> error.toString
+        )
```

Comment:
Make this a `Writes`, also I don't think this is correct, the `type` should be `error` if this is an error message, and the payload should be an error object not a string.

Reply:
I've made Writes and Reads for PlayWSMessage,  the code has been simplified, and it seems more idiomatic. Thank you for the tip!

Reply:
I'm not sure if the `type` should be `error` here. I've checked the other adapters, and they use `complete` in this case. Maybe we need to check the protocol, but I haven't found a lot of docs about it (only https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md) Do you have more info or any doc @ghostdogpr?

Reply:
Yeah the docs are not very detailed in this case. Looking at the code, I think this can never happen actually. We have a `Throwable` because we wrap Futures but the only thing they do is offering to a queue. I think you can keep it consistent with the other adapters and we might clean that up later.